# Introduction

This project analyzes customer shopping behavior for a leading retail company. 
The company wants to understand factors influencing purchases, loyalty, and subscription adoption to improve sales, marketing strategies, and customer engagement.

Business Problem:
The management observed changes in purchasing patterns across demographics, product categories, and sales channels. 
Objective: Identify trends, high-value customers, top-selling products, and factors driving purchases and loyalty.

Dataset Overview:
The dataset contains 3900 customer transaction records with 18 columns, including demographic data, purchase details, product information, subscription status, and previous purchases.

Scope & Approach:
- Data Cleaning & Feature Engineering using Python (Pandas)
- Exploratory Data Analysis using SQL queries
- Final data export for Power BI visualization
- Insights & Recommendations for business decision-making


# Importing Libraries

In [44]:
# Data processing
import numpy as np
import pandas as pd

%matplotlib inline

# Display formatting
from IPython.core.display import HTML
HTML("""
<style>
.dataframe table, .dataframe th, .dataframe td { font-size: 12px; }
div.output_scroll { overflow-x: auto; }
</style>
""")

# Loading Dataset

In [2]:
df = pd.read_csv("customer_shopping_behavior.csv")
df

Customer ID  Age  Gender Item Purchased     Category  \
0               1   55    Male         Blouse     Clothing   
1               2   19    Male        Sweater     Clothing   
2               3   50    Male          Jeans     Clothing   
3               4   21    Male        Sandals     Footwear   
4               5   45    Male         Blouse     Clothing   
...           ...  ...     ...            ...          ...   
3895         3896   40  Female         Hoodie     Clothing   
3896         3897   52  Female       Backpack  Accessories   
3897         3898   46  Female           Belt  Accessories   
3898         3899   44  Female          Shoes     Footwear   
3899         3900   52  Female        Handbag  Accessories   

      Purchase Amount (USD)       Location Size      Color  Season  \
0                        53       Kentucky    L       Gray  Winter   
1                        64          Maine    L     Maroon  Winter   
2                        73  Massachusetts    S     Maroon  Spring   
3                        90   Rhode Island    M     Maroon  Spring   
4                        49         Oregon    M  Turquoise  Spring   
...                     ...            ...  ...        ...     ...   
3895                     28       Virginia    L  Turquoise  Summer   
3896                     49           Iowa    L      White  Spring   
3897                     33     New Jersey    L      Green  Spring   
3898                     77      Minnesota    S      Brown  Summer   
3899                     81     California    M      Beige  Spring   

      Review Rating Subscription Status   Shipping Type Discount Applied  \
0               3.1                 Yes         Express              Yes   
1               3.1                 Yes         Express              Yes   
2               3.1                 Yes   Free Shipping              Yes   
3               3.5                 Yes    Next Day Air              Yes   
4               2.7                 Yes   Free Shipping              Yes   
...             ...                 ...             ...              ...   
3895            4.2                  No  2-Day Shipping               No   
3896            4.5                  No    Store Pickup               No   
3897            2.9                  No        Standard               No   
3898            3.8                  No         Express               No   
3899            3.1                  No    Store Pickup               No   

     Promo Code Used  Previous Purchases Payment Method Frequency of Purchases  
0                Yes                  14          Venmo            Fortnightly  
1                Yes                   2           Cash            Fortnightly  
2                Yes                  23    Credit Card                 Weekly  
3                Yes                  49         PayPal                 Weekly  
4                Yes                  31         PayPal               Annually  
...              ...                 ...            ...                    ...  
3895              No                  32          Venmo                 Weekly  
3896              No                  41  Bank Transfer              Bi-Weekly  
3897              No                  24          Venmo              Quarterly  
3898              No                  24          Venmo                 Weekly  
3899              No                  33          Venmo              Quarterly  

[3900 rows x 18 columns]

### Initial Impression
 - Check dataset info.
 - Identify missing values or inconsistencies.
 - Generate statistical summaries using `df.describe()` to review the distribution and basic statistics of all.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   object 
 3   Item Purchased          3900 non-null   object 
 4   Category                3900 non-null   object 
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   object 
 7   Size                    3900 non-null   object 
 8   Color                   3900 non-null   object 
 9   Season                  3900 non-null   object 
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   object 
 12  Shipping Type           3900 non-null   object 
 13  Discount Applied        3900 non-null   object 
 14  Promo Code Used         3900 non-null   

In [4]:
df.describe().round(2)

Customer ID      Age  Purchase Amount (USD)  Review Rating  \
count      3900.00  3900.00                3900.00        3863.00   
mean       1950.50    44.07                  59.76           3.75   
std        1125.98    15.21                  23.69           0.72   
min           1.00    18.00                  20.00           2.50   
25%         975.75    31.00                  39.00           3.10   
50%        1950.50    44.00                  60.00           3.80   
75%        2925.25    57.00                  81.00           4.40   
max        3900.00    70.00                 100.00           5.00   

       Previous Purchases  
count             3900.00  
mean                25.35  
std                 14.45  
min                  1.00  
25%                 13.00  
50%                 25.00  
75%                 38.00  
max                 50.00

In [5]:
df.describe(include='O')

Gender Item Purchased  Category Location  Size  Color  Season  \
count    3900           3900      3900     3900  3900   3900    3900   
unique      2             25         4       50     4     25       4   
top      Male         Blouse  Clothing  Montana     M  Olive  Spring   
freq     2652            171      1737       96  1755    177     999   

       Subscription Status  Shipping Type Discount Applied Promo Code Used  \
count                 3900           3900             3900            3900   
unique                   2              6                2               2   
top                     No  Free Shipping               No              No   
freq                  2847            675             2223            2223   

       Payment Method Frequency of Purchases  
count            3900                   3900  
unique              6                      7  
top            PayPal         Every 3 Months  
freq              677                    584

# Data Cleaning & Preparation

In [6]:
df.columns

Index(['Customer ID', 'Age', 'Gender', 'Item Purchased', 'Category',
       'Purchase Amount (USD)', 'Location', 'Size', 'Color', 'Season',
       'Review Rating', 'Subscription Status', 'Shipping Type',
       'Discount Applied', 'Promo Code Used', 'Previous Purchases',
       'Payment Method', 'Frequency of Purchases'],
      dtype='object')

In [7]:
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ','_')
df = df.rename(columns={'purchase_amount_(usd)':'purchase_amount'})

In [8]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'promo_code_used', 'previous_purchases',
       'payment_method', 'frequency_of_purchases'],
      dtype='object')

**Note on Column Renaming**

We renamed the dataset columns to **follow a consistent and professional format** (lowercase with underscores).  
This makes the dataset easier to work with in Python and SQL, especially for:

- Writing clean and readable code
- Avoiding errors caused by spaces or special characters in column names

In [9]:
df.isna().sum()

customer_id                0
age                        0
gender                     0
item_purchased             0
category                   0
purchase_amount            0
location                   0
size                       0
color                      0
season                     0
review_rating             37
subscription_status        0
shipping_type              0
discount_applied           0
promo_code_used            0
previous_purchases         0
payment_method             0
frequency_of_purchases     0
dtype: int64

In [10]:
df['review_rating'] = df.groupby('category')['review_rating']\
                         .transform(lambda x: x.fillna(x.median()))

**Reasoning:**
Review ratings were filled using the median value per category to maintain consistency within product groups and avoid bias caused by outliers.

In [11]:
df.duplicated().sum()

np.int64(0)

# Feature Engineering

In [12]:
df.sample(5)

customer_id  age  gender item_purchased     category  purchase_amount  \
3884         3885   47  Female        Sandals     Footwear               29   
1074         1075   48    Male        Jewelry  Accessories               71   
1522         1523   38    Male         Shorts     Clothing               72   
2042         2043   19    Male          Shoes     Footwear               52   
2945         2946   32  Female           Belt  Accessories               49   

       location size     color  season  review_rating subscription_status  \
3884   Maryland    M     Brown    Fall            4.9                  No   
1074  Tennessee    S     White    Fall            2.6                  No   
1522  Tennessee    L    Yellow  Spring            4.8                  No   
2042    Alabama    M       Red  Spring            4.0                  No   
2945   Delaware    M  Charcoal  Summer            4.8                  No   

       shipping_type discount_applied promo_code_used  previous_purchases  \
3884         Express               No              No                   3   
1074        Standard              Yes             Yes                  45   
1522   Free Shipping              Yes             Yes                   5   
2042  2-Day Shipping               No              No                  35   
2945   Free Shipping               No              No                   1   

     payment_method frequency_of_purchases  
3884         PayPal                 Weekly  
1074  Bank Transfer            Fortnightly  
1522     Debit Card                Monthly  
2042          Venmo                Monthly  
2945    Credit Card         Every 3 Months

In [13]:
bins = [17, 25, 40, 55, 100]
labels = ['young_adult', 'adult', 'middle_aged', 'senior']

df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels)

In [14]:
df['age_group'].value_counts()


age_group
middle_aged    1121
senior         1105
adult          1103
young_adult     571
Name: count, dtype: int64

**Note:**
The age_group feature was created to segment customers into meaningful life-stage categories.
This enables clearer analysis of purchasing behavior across different age segments and supports targeted marketing and customer engagement strategies.

In [15]:
frequency_map = {
    'Weekly': 7,
    'Bi-Weekly': 14,
    'Fortnightly': 14,
    'Monthly': 30,
    'Every 3 Months': 90,
    'Quarterly': 90,
    'Annually': 365
}

df['frequency_of_purchases_days'] = df['frequency_of_purchases'].map(frequency_map)

In [16]:
df[['frequency_of_purchases', 'frequency_of_purchases_days']].head()

frequency_of_purchases  frequency_of_purchases_days
0            Fortnightly                           14
1            Fortnightly                           14
2                 Weekly                            7
3                 Weekly                            7
4               Annually                          365

**Note:**
`The frequency_of_purchases_days` feature was created by mapping textual purchase frequency labels into approximate numerical values (in days).
This transformation allows for quantitative analysis, easier comparisons, and correlation with other numerical variables such as purchase amount and customer loyalty.

In [17]:
(df['discount_applied'] == df['promo_code_used']).all()

np.True_

In [18]:
df = df.drop(columns=['promo_code_used'])

In [19]:
df.columns

Index(['customer_id', 'age', 'gender', 'item_purchased', 'category',
       'purchase_amount', 'location', 'size', 'color', 'season',
       'review_rating', 'subscription_status', 'shipping_type',
       'discount_applied', 'previous_purchases', 'payment_method',
       'frequency_of_purchases', 'age_group', 'frequency_of_purchases_days'],
      dtype='object')

**Note:**
The `promo_code_used` column was removed after confirming that it contains identical information to the `discount_applied` column across all records.
To avoid redundancy and simplify analysis, only `discount_applied` was retained as it more directly represents the business outcome.

# EDA using SQL

In [20]:
from sqlalchemy import create_engine

engine = create_engine("sqlite:///retail.db")

df.to_sql(
    "customer_transactions",
    engine,
    if_exists="replace",
    index=False
)

3900

In [21]:
pd.read_sql(
    "SELECT COUNT(*) AS total_rows FROM customer_transactions",
    engine
)

total_rows
0        3900

**Note:**
“I used **SQLAlchemy** as a database abstraction layer to load cleaned data into a relational database and perform SQL-based analysis, ensuring portability across database systems.”

In [23]:
query = """
SELECT
    gender,
    COUNT(DISTINCT customer_id) AS customers_count,
    SUM(purchase_amount) AS total_revenue,
    AVG(purchase_amount) AS avg_purchase_amount
FROM customer_transactions
GROUP BY gender;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

gender  customers_count  total_revenue  avg_purchase_amount
0  Female             1248          75191            60.249199
1    Male             2652         157890            59.536199

**Business Insight:**
While male customers generate higher total revenue, this is driven by volume rather than higher spending per transaction. 
Female customers demonstrate comparable average purchase values, indicating strong revenue potential if customer acquisition efforts are expanded toward female segments.

**Recommendation:**
Focus marketing and acquisition campaigns on expanding the female customer base, as their spending behavior matches that of male customers and presents an opportunity for revenue growth.


In [24]:
query = """
SELECT
    COUNT(*) AS high_value_discount_customers,
    SUM(purchase_amount) AS total_revenue
FROM customer_transactions
WHERE discount_applied = 'Yes'
  AND purchase_amount > (
      SELECT AVG(purchase_amount)
      FROM customer_transactions
  );

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

high_value_discount_customers  total_revenue
0                            839          66942

**Insight:**
Approximately 839 customers (around 21.5% of the customer base) used discounts while still spending above the average purchase amount, generating nearly $67K in revenue. 
This indicates a significant segment of high-value customers who are not strongly price-sensitive.

**Recommendation:**
Instead of offering standard discounts to this segment, the company should consider shifting toward value-based incentives such as loyalty rewards, exclusive access, or bundled offers to maintain high revenue while protecting profit margins.

In [25]:
query = """
SELECT
    AVG(previous_purchases) AS avg_loyalty
FROM customer_transactions
WHERE discount_applied = 'Yes'
  AND purchase_amount > (
      SELECT AVG(purchase_amount)
      FROM customer_transactions
  );

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

avg_loyalty
0    25.697259

**Insight:**
High-spending customers who used discounts show an average of approximately 25.7 previous purchases, which is slightly above the overall customer average. 
This indicates that this segment is not only high-value in terms of spending but also demonstrates strong loyalty and repeat purchasing behavior.

**Recommendation:**
Rather than removing discounts entirely for this segment, the company should redesign discount strategies to reward loyalty, such as tier-based discounts, personalized offers, or exclusive benefits for repeat customers, ensuring both customer retention and margin optimization.

In [30]:
query = """
SELECT
    item_purchased,
    AVG(review_rating) AS avg_rating,
    COUNT(review_rating) AS review_count
FROM customer_transactions
GROUP BY item_purchased
HAVING COUNT(review_rating) >= 20
ORDER BY avg_rating DESC
LIMIT 5;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

item_purchased  avg_rating  review_count
0         Gloves    3.861429           140
1        Sandals    3.844375           160
2          Boots    3.818750           144
3            Hat    3.801299           154
4          Skirt    3.784810           158

**Insight:**
The top five products by average review rating show consistently similar ratings, all ranging between approximately 3.78 and 3.86, with a substantial number of reviews for each. 
This indicates a stable level of customer satisfaction across multiple product types rather than a single standout product.

**Recommendation:**
Since no single product significantly outperforms others in customer ratings, the company should focus on incremental improvements across its top product categories, such as enhancing product features, sizing accuracy, or post-purchase support, rather than relying on a single flagship item.


In [31]:
query = """
SELECT
    item_purchased,
    COUNT(*) AS total_sales,
    AVG(review_rating) AS avg_rating
FROM customer_transactions
GROUP BY item_purchased
ORDER BY total_sales DESC
LIMIT 5;


"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

item_purchased  total_sales  avg_rating
0          Pants          171    3.720468
1        Jewelry          171    3.759064
2         Blouse          171    3.680702
3          Shirt          169    3.622485
4          Dress          166    3.749398

**Insight:**
The company’s top-selling products do not overlap with its highest-rated products. 
While items like Pants and Blouses generate the highest sales volume, their average review ratings are relatively moderate compared to other product categories.
This suggests that purchase decisions are driven more by necessity, fashion trends, or pricing rather than perceived product quality.

**Recommendation:**
The company should prioritize quality and customer experience improvements for its top-selling products. 
Even small enhancements in fit, material quality, or sizing consistency for high-volume items like Pants and Shirts could significantly boost overall customer satisfaction and repeat purchases.


In [32]:
query = """
SELECT
    item_purchased,
    AVG(previous_purchases) AS avg_repeat_purchases,
    AVG(review_rating) AS avg_rating
FROM customer_transactions
GROUP BY item_purchased
ORDER BY avg_repeat_purchases DESC
LIMIT 5;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

item_purchased  avg_repeat_purchases  avg_rating
0        Jewelry             28.906433    3.759064
1           Coat             26.813665    3.728571
2          Scarf             26.783439    3.707006
3         Blouse             26.684211    3.680702
4          Dress             26.548193    3.749398

**Insight:**
Products with the highest repeat purchase rates do not necessarily have the highest customer review ratings.
Items such as Jewelry and Coats show strong customer loyalty despite receiving only moderate average ratings.
This indicates that repeat purchases are likely driven by factors such as product necessity, variety, pricing, or brand trust rather than review scores alone.

**Recommendation:**
The company should focus on improving the customer experience of its high-loyalty products.
Since customers are already repurchasing these items frequently, targeted improvements in quality, fit, or durability could significantly increase satisfaction and lifetime value without needing aggressive marketing campaigns.


In [33]:
query = """
SELECT
    gender,
    age_group,
    AVG(previous_purchases) AS avg_loyalty,
    AVG(purchase_amount) AS avg_spending
FROM customer_transactions
GROUP BY gender, age_group
ORDER BY avg_loyalty DESC;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

gender    age_group  avg_loyalty  avg_spending
0    Male       senior    26.643890     58.467806
1    Male  middle_aged    25.885180     60.372497
2  Female        adult    25.401114     60.610028
3  Female       senior    25.194767     60.354651
4    Male  young_adult    25.148241     60.537688
5    Male        adult    24.868280     59.251344
6  Female  middle_aged    23.798387     59.500000
7  Female  young_adult    23.450867     60.901734

**Insight:**
Customer loyalty varies significantly across demographic segments, while average spending per purchase remains relatively consistent.
Male senior and middle-aged customers exhibit the highest repeat purchase behavior, indicating stronger long-term engagement.
In contrast, younger female customers show lower loyalty despite similar spending levels, suggesting higher churn risk.

**Recommendation:**
The company should adopt a segmented marketing strategy.
High-loyalty male senior customers should be targeted with retention-focused campaigns, while younger female customers require engagement-driven initiatives to reduce churn and increase repeat purchases.


In [34]:
query = """
SELECT
    shipping_type,
    COUNT(*) AS total_orders,
    AVG(purchase_amount) AS avg_purchase_amount
FROM customer_transactions
GROUP BY shipping_type;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

shipping_type  total_orders  avg_purchase_amount
0  2-Day Shipping           627            60.733652
1         Express           646            60.475232
2   Free Shipping           675            60.410370
3    Next Day Air           648            58.631173
4        Standard           654            58.460245
5    Store Pickup           650            59.893846

**Insight:**
Average purchase amounts are relatively consistent across different shipping types.
Customers who choose faster shipping options do not significantly spend more than those using standard or free shipping.
This suggests that shipping preference is driven primarily by convenience rather than customer spending power.

**Recommendation:**
The company should position shipping options as a convenience and experience feature rather than a revenue-driving lever.
Investments in faster or flexible shipping should focus on customer satisfaction and retention rather than increasing average order value.


In [35]:
query = """
SELECT
    subscription_status,
    COUNT(DISTINCT customer_id) AS customers_count,
    AVG(purchase_amount) AS avg_purchase_amount,
    SUM(purchase_amount) AS total_revenue
FROM customer_transactions
GROUP BY subscription_status;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

subscription_status  customers_count  avg_purchase_amount  total_revenue
0                  No             2847            59.865121         170436
1                 Yes             1053            59.491928          62645

**Insight:**
Subscribed customers spend roughly the same per purchase as non-subscribers, with an average of about `$59.5` compared to `$59.9`. 
However, due to their smaller numbers, the total revenue from subscribers is significantly lower. 
This indicates that subscriptions do not currently drive higher individual spending.

**Recommendation:**
To increase the impact of subscriptions, the company should consider introducing incentives that encourage higher spending among subscribers, such as exclusive offers, bundled deals, or loyalty rewards.
Alternatively, marketing efforts could focus on converting high-value non-subscribers into subscribers to balance revenue and engagement.


In [36]:
query = """
SELECT
    item_purchased,
    COUNT(*) AS total_orders,
    SUM(CASE WHEN discount_applied = 'Yes' THEN 1 ELSE 0 END) AS discount_orders,
    ROUND(100.0 * SUM(CASE WHEN discount_applied = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS discount_percentage
FROM customer_transactions
GROUP BY item_purchased
ORDER BY discount_percentage DESC
LIMIT 5;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

item_purchased  total_orders  discount_orders  discount_percentage
0            Hat           154               77                50.00
1       Sneakers           145               72                49.66
2           Coat           161               79                49.07
3        Sweater           164               79                48.17
4          Pants           171               81                47.37

**Insight:**
Products such as Hats, Sneakers, Coats, Sweaters, and Pants have nearly half of their purchases made with discounts applied. 
This indicates that these items are highly responsive to promotions and that discounts play a significant role in driving sales for these categories.

**Recommendation:**
The company should consider optimizing discount strategies for these high-discount products. 
Options include targeted promotions, bundling with high-margin items, or introducing loyalty incentives to maintain sales without eroding profit margins.


In [37]:
query = """
SELECT
    CASE
        WHEN previous_purchases <= 5 THEN 'New'
        WHEN previous_purchases BETWEEN 6 AND 20 THEN 'Returning'
        ELSE 'Loyal'
    END AS customer_segment,
    COUNT(DISTINCT customer_id) AS customers_count
FROM customer_transactions
GROUP BY customer_segment
ORDER BY customers_count DESC;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

customer_segment  customers_count
0            Loyal             2339
1        Returning             1137
2              New              424

**Insight:**
The majority of customers are classified as Loyal, indicating strong repeat purchase behavior in the customer base.
Returning customers represent a significant portion, providing opportunities to increase loyalty.
The relatively small number of New customers highlights the importance of customer acquisition initiatives.

**Recommendation:**
1. Implement loyalty programs and exclusive rewards to retain Loyal customers.
2. Offer targeted promotions or personalized recommendations to Returning customers to encourage repeat purchases.
3. Develop acquisition campaigns to increase the number of New customers, ensuring sustainable growth of the customer base.


In [39]:
query = """
WITH product_sales AS (
    SELECT
        category,
        item_purchased,
        COUNT(*) AS total_orders
    FROM customer_transactions
    GROUP BY category, item_purchased
)
SELECT
    category,
    item_purchased,
    total_orders
FROM (
    SELECT
        category,
        item_purchased,
        total_orders,
        ROW_NUMBER() OVER (PARTITION BY category ORDER BY total_orders DESC) AS rn
    FROM product_sales
) t
WHERE rn <= 3
ORDER BY category, total_orders DESC;


"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

category item_purchased  total_orders
0   Accessories        Jewelry           171
1   Accessories           Belt           161
2   Accessories     Sunglasses           161
3      Clothing         Blouse           171
4      Clothing          Pants           171
5      Clothing          Shirt           169
6      Footwear        Sandals           160
7      Footwear          Shoes           150
8      Footwear       Sneakers           145
9     Outerwear         Jacket           163
10    Outerwear           Coat           161

**Insight:**
The top 3 products within each category are the primary drivers of sales.
Accessories are led by Jewelry, Belt, and Sunglasses, while Clothing is dominated by Blouse, Pants, and Shirt.
Footwear and Outerwear show similar patterns with a few high-demand items driving the majority of category sales.

**Recommendation:**
1. Prioritize inventory management for the top-selling products to avoid stockouts.
2. Highlight these products in marketing campaigns as key offerings.
3. Consider bundling top products with slower-selling items to increase overall category revenue.


In [40]:
query = """
SELECT
    CASE 
        WHEN previous_purchases > 5 THEN 'Repeat Buyer'
        ELSE 'Occasional Buyer'
    END AS buyer_type,
    COUNT(*) AS total_customers,
    SUM(CASE WHEN subscription_status = 'Yes' THEN 1 ELSE 0 END) AS subscribers_count,
    ROUND(100.0 * SUM(CASE WHEN subscription_status = 'Yes' THEN 1 ELSE 0 END) / COUNT(*), 2) AS subscriber_percentage
FROM customer_transactions
GROUP BY buyer_type;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

buyer_type  total_customers  subscribers_count  subscriber_percentage
0  Occasional Buyer              424                 95                  22.41
1      Repeat Buyer             3476                958                  27.56

**Insight:**
Repeat buyers are slightly more likely to be subscribed compared to occasional buyers, with 27.56% of repeat buyers having a subscription versus 22.41% of occasional buyers.
This indicates a modest correlation between customer loyalty and subscription adoption.

**Recommendation:**
1. Focus marketing campaigns for subscription upsell on repeat buyers, as they are more likely to adopt.
2. Consider targeted incentives for occasional buyers to encourage both repeat purchases and subscription adoption.
3. Monitor subscription adoption trends to identify if these efforts increase overall customer loyalty.


In [41]:
query = """
SELECT
    age_group,
    COUNT(*) AS total_orders,
    SUM(purchase_amount) AS total_revenue,
    AVG(purchase_amount) AS avg_purchase_amount
FROM customer_transactions
GROUP BY age_group
ORDER BY total_revenue DESC;

"""

revenue_by_gender = pd.read_sql(query, engine)
revenue_by_gender

age_group  total_orders  total_revenue  avg_purchase_amount
0  middle_aged          1121          67353            60.082962
1        adult          1103          65842            59.693563
2       senior          1105          65256            59.055204
3  young_adult           571          34630            60.647986

**Insight:**
The middle-aged segment contributes the most to total revenue, followed closely by adult and senior segments.
Young adults, while having the highest average purchase amount per order, contribute the least to overall revenue due to fewer orders.

**Recommendation:**
1. Focus marketing campaigns and promotions on middle-aged and adult customers to maintain high revenue contribution.
2. Develop targeted campaigns for young adults to increase order frequency and engagement.
3. Leverage insights from average purchase amounts to optimize pricing, bundles, and promotions per age group.


# Data Cleaning, Feature Selection, and Database/CSV Export

In [42]:
final_columns = [
    'customer_id', 'age', 'age_group', 'gender', 'item_purchased', 
    'category', 'purchase_amount', 'location', 'size', 'color', 'season', 
    'review_rating', 'subscription_status', 'shipping_type', 
    'discount_applied', 'previous_purchases', 'payment_method', 
    'frequency_of_purchases_days'
]

final_df = df[final_columns]

final_df.to_csv("customer_transactions_final.csv", index=False)

In [43]:
from sqlalchemy import create_engine

engine = create_engine('sqlite:///retail_final.db')

final_df.to_sql("customer_transactions_final", engine, if_exists='replace', index=False)


3900

**Note:**
- Selected the final set of relevant columns for analysis and removed temporary or redundant features.
- Ensured all data is clean, consistent, and properly formatted.
- Saved a finalized CSV file ('customer_transactions_final.csv') for use in Power BI.
- Created a finalized SQLite database ('retail_final.db') with the table 'customer_transactions_final' for professional database connection and direct analysis in Power BI.
